### DISCUSSION TITLE
Debug Console

### DISCUSSION THREAD

**Discussion by thatbordercollie** [ikemen-engine/Ikemen-GO#1488] [Engine development]
I use mugen 1.1's control+tilde debug console a lot to govern the life of AI players in simul - I wonder if it could be ported to ikemen. 

If it was to be ported over, it would share commands with mugen (eg. `setlife 1 1000`) while adding ikemen-exclusive commands - for example maybe you can play a specific sound with `playsound path/to/sound.wav`?

**Comment by ZyonZZ** [comment]
in IKEMEN it is Ctrl-D, but it only shows information, you cannot enter commands themselves...
and they are the same as Ctrl-C for the collision boxes

I don't know what you mean by "govern life"
but
  F1...kill 2p
Ctrl-F1...kill 1p

ctrl-1 = activate/deactivate AI p1
ctrl-2 = activate/deactivate AI p2
ctrl-3 = activate/deactivate AI p3
ctrl-4 = activate/deactivate AI p4

F2 low life of both

F3 max power bar
f4 reset round
Shift-f4 Reload round (all archives:: chars, fight, stages... I use this for when I am editing something in a char, stage, lifebar etc)
and I'm sure there are more that I don't know about... they're probably in the "Ikemen Wiki"
And if you were looking for a debug for internal processes...

which you open in CMD with this command ""ikemen_go.exe run --debug""
You can see a few processes... but not everything...

**Comment by Lazin3ss** [comment]
This is kind of a feature port (since I don't think we'll do the same implementation as MUGEN in case we develop it), and as such, it should go into issues page or feature request.
@K4thos I leave it to your consideration

**Comment by Jesuszilla** [comment]
Bumping this up because I was actually discussing this in the dev room in Discord not too long ago. On lunch break so I figured I might as well post my ideas now.

So I don't think replicating the MConsole exactly as it was is a good idea mainly because of how much Lua this engine is capable of. So instead of replicating exact MConsole functionality, my proposal instead is to add a console that is capable of executing arbitary Lua commands. It could either be called the lconsole for Lua console, or iconsole for interactive console.

I intend to start implementation soon, as it appears to mostly be Lua work and I probably won't have to touch _too much_ Go, but I know I'll have to touch some mainly to register a new feature to go along with the console: a `man` (manual) command, because `help` like MConsole had is certainly not feasible with all the Lua commands this engine has registered currently. The console would allow users to search for commands by name, as well as read a description of what a given command does and display their allowed arguments and their types.

examples of new methods to go with this console:

`man("animAddPos")` would output the following into the console:
```
description:
Offsets the given anim by the given x,y coordinates.

args:
anim (Anim) - the anim to use
x (int) - the x position to offset by
y (int) - the y position to offset by
```

`findLuaMethod("bg")` might output the following (there's more I've left out for brevity) into the console:

```
Method signatures matching provided text: bgDraw, bgmLength, bgmPosition, bgNew, bgReset, playBGM, setVolumeBgm
```

I've also already captured a list of the MConsole commands that currently have Lua equivalents in I.K.E.M.E.N-Go, as well as new functions that would need to be added to replicate the full functionality of MConsole [here](https://pastebin.com/Qgrp3csw).

Additionally, I'll see if I can provide a way to search for entities to move the debug cursor to a specified entity using any of the following proposed new functions: `findEntityByName`, `findEntityByPlayerId`, and `findHelperById`, which would move the debug console display to the next found entity matching the criteria.

So to reiterate, I believe the console shouldn't be exactly the same as what M.U.G.E.N does on the basis of that:
1. 1.1 was never really completed
2. Elecbyte seemed to be moving in a direction where Lua would play a bigger role in the engine that I.K.E.M.E.N-Go has outdone by this point.
3. This engine has so much Lua functionality that I don't recommend gimping what the console could do just to replicate a weaker console that doesn't even stick with its own naming conventions for the feature (they have toggleai, togglelifebars, and toggleplayer, but toggling the console is called consoletoggle for instance)
4. It's debug and shouldn't affect too many legacy users because of how limited the MConsole was in M.U.G.E.N

TL;DR why replicate MConsole exactly when something better can definitely be made

EDIT: I got a basic display going and it reads and executes Lua commands now!

![ikemen013](https://github.com/ikemen-engine/Ikemen-GO/assets/20290325/3b9c6b6b-4bd2-4096-9c77-f732e71b82b1)
![ikemen014](https://github.com/ikemen-engine/Ikemen-GO/assets/20290325/86a8f1fb-910d-40d7-b5a4-52d041817ae9)

EDIT2: History and arrow navigation working, and `man` partially implemented:
![ikemen021](https://github.com/ikemen-engine/Ikemen-GO/assets/20290325/465154e3-388c-4425-bc67-7e49becd2c84)

EDIT3: man and findLuaMethod refactored into different commands, Lua module creators can now access a global table called `man_table` or just use the `register_man` convenience method to register their own commands with `man`:
![ikemen032](https://github.com/ikemen-engine/Ikemen-GO/assets/20290325/120d23f2-007f-4931-841c-b813b786bab3)

**Reply by Jesuszilla** [reply]
Little update on this: implemented a way to print to the interactive console with `iconsole.printString(str)` which prints a string into the interactive console, taking into account any necessary word wrapping or newlines. Search functionality has been implemented as well:

![ikemen044](https://github.com/ikemen-engine/Ikemen-GO/assets/20290325/d3066539-1081-421f-9a81-bda278cc3e0d)
![ikemen046](https://github.com/ikemen-engine/Ikemen-GO/assets/20290325/f9aaf18a-c420-49d2-a475-dae256befa31)
![ikemen047](https://github.com/ikemen-engine/Ikemen-GO/assets/20290325/a0f273be-b94b-41a2-be01-eb05eaa22330)
